jax_distribution/logger.py

```diff
@@ -0,0 +1,136 @@
+
+import chex
+import jax.numpy as jnp
+import numpy as np
+from sacred.run import Run
+
+from jax_distribution.utils.logger_tools import Logger
+
+
+def get_logger_fn(logger: SacredLogger, config: Dict) -> Callable:
+    """Get the logger function."""
+
+    def log(
+        metrics: Dict[str, Dict[str, chex.Array]],
+        t_env: int = 0,
+        trainer_metric: bool = False,
+        absolute_metric: bool = False,
+    ) -> None:
+        """Log the episode returns and lengths.
+
+        Args:
+            metrics (Dict): The metrics info.
+            t_env (int): The current timestep.
+            trainer_metric (bool): Whether to log the trainer metric.
+            absolute_metric (bool): Whether to log the absolute metric.
+        """
+        if absolute_metric:
+            prefix = "Absolute_"
+            metrics_info = metrics
+
+        elif trainer_metric:
+            prefix = "Trainer_"
+            metrics_info = metrics["metrics"]
+            total_loss = metrics["total_loss"]
+            value_loss = metrics["value_loss"]
+            loss_actor = metrics["loss_actor"]
+            entropy = metrics["entropy"]
+        else:
+            prefix = ""
+            metrics_info = metrics
+
+        # Flatten metrics info.
+        episodes_return = jnp.ravel(metrics_info["episode_return"])
+        episodes_length = jnp.ravel(metrics_info["episode_length"])
+
+        # Log metrics.
+        if config["USE_SACRED"] or config["USE_TF"]:
+            logger.log_stat(
+                prefix.lower() + "mean_test_episode_returns",
+                float(np.mean(episodes_return)),
+                t_env,
+            )
+            logger.log_stat(
+                prefix.lower() + "mean_test_episode_length",
+                float(np.mean(episodes_length)),
+                t_env,
+            )
+            if trainer_metric:
+                logger.log_stat("total_loss", float(np.mean(total_loss)), t_env)
+                logger.log_stat("value_loss", float(np.mean(value_loss)), t_env)
+                logger.log_stat("loss_actor", float(np.mean(loss_actor)), t_env)
+                logger.log_stat("entropy", float(np.mean(entropy)), t_env)
+
+        log_string = "Timesteps {:07d}".format(t_env) + " "
+        log_string += "| Mean Episode Returns {:.3f} ".format(
+            float(np.mean(episodes_return))
+        )
+        log_string += "| Std Episode Returns {:.3f} ".format(
+            float(np.std(episodes_return))
+        )
+        log_string += "| Max Episode Returns {:.3f} ".format(
+            float(np.max(episodes_return))
+        )
+        log_string += "| Mean Episode Length {:.3f} ".format(
+            float(np.mean(episodes_length))
+        )
+        log_string += "| Std Episode Length {:.3f} ".format(
+            float(np.std(episodes_length))
+        )
+        log_string += "| Max Episode Length {:.3f} ".format(
+            float(np.max(episodes_length))
+        )
```

Comment:
This is because we wanted to have round the values by having 3 after the "," however we can change it to f"string" if you want and we use by that the `round` function to achieve what we wanted

Reply:
I updated it to look like this: 
```
log_string = (
            "Timesteps {:07d} | "
            "Mean Episode Return {:.3f} | "
            "Std Episode Return {:.3f} | "
            "Max Episode Return {:.3f} | "
            "Mean Episode Length {:.3f} | "
            "Std Episode Length {:.3f} | "
            "Max Episode Length {:.3f}"
        ).format(
            t_env,
            float(np.mean(episodes_return)),
            float(np.std(episodes_return)),
            float(np.max(episodes_return)),
            float(np.mean(episodes_length)),
            float(np.std(episodes_length)),
            float(np.max(episodes_length))
        )
```
I don't know if this solved the issue of "+=" because using """string""" causing some spacing issues in the terminal output

Reply:
Would this work? I just think `f"string"` is much easier to read

```python
log_string = (
            f"Timesteps {t_env:07d} | "
            f"Mean Episode Return {float(np.mean(episodes_return)):.3f} | "
            f"Std Episode Return {float(np.std(episodes_return)):.3f} | "
            f"Max Episode Return {float(np.max(episodes_return)):.3f} | "
            f"Mean Episode Length {float(np.mean(episodes_length)):.3f} | "
            f"Std Episode Length {float(np.std(episodes_length)):.3f} | "
            f"Max Episode Length {float(np.max(episodes_length)):.3f}"
        )
```